doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,530 @@
+
+<img src="images/heartbeat.png" alt="heartbeat" width="500px"/>
+
+#### 1.2.1.8 Authentication
+sonic-telemetry provides three authentication mechanisms, and sonic-gnmi will use the same mechanisms:
+* Password: Like HTTP Basic auth, you pass the username and password in the gRPC metadata
+* JWT: JSON Web Tokens standard. First you authenticate with username/password and then receive a JWT token. After you send the token in the gRPC metadata.
+* User Certificate: Use a valid client certificate with the username embedded in the CN field. The certificate is for authentication, the username is for authorization.
+
+sonic-restapi uses one authentication mechanism:
+* In the authentication of client certificate, after the certificate chain is validated by TLS, it will further check if the common name of the end-entity certificate is in the trusted common name list of the server config.
+
+#### 1.2.1.9 ACL
+GNMI server will not support OpenConfig Yang models, so ACL configuration will use CONFIG_DB schema and SONiC Yang Models.
+
+#### 1.2.1.10 Docker to Host communication
+'config apply-patch' and 'config reload' are designed to run on host, and it's difficult to support them in container:
+1. These commands will update redis database and restart container, when they restart gnmi, bgp, syncd and swss, the ongoing gnmi operation will be broken.
+2. 'config reload' will stop service at first, run some other operations, and then restart service. If we run this command in container, it will be broken at the stop service step.
+3. These commands will execute some host scripts and use systemctl to restart service, it will be dangerous to support these operations in container.
+The solution is to add host services for 'config apply-patch' and 'config reload' on host, and gNMI server uses dbus method to invoke these services to update configuration
+
+##### 1.2.1.10.1 Incremental Configurations
+
+For incremental configuration, 'config apply-patch' should not restart gNMI, bgp, synd and swss.
+
+<img src="images/incremental_rpc.png" alt="incremental" width="500px"/>
+
+##### 1.2.1.10.2 Full Configurations
+
+For full configuration, there’re 2 possible solutions: 
+* gNMI server needs to send response at first, and then invoke 'config reload'.
+
+<img src="images/full_rpc.png" alt="full-gnmi" width="500px"/>
+
+* Use gNMI request and gNOI request together to implement full configuration update.
+
+<img src="images/full_rpc_gnoi.png" alt="full-gnmi-gnoi" width="500px"/>
+
+The full configuration request will be overwritten by subsequent full configuration request or incremental configuration request.
+
+<img src="images/mixed requests.png" alt="overwritten-config" width="500px"/>
+
+### 1.2.2 Container
+
+All the introduced features will be part of the sonic-gnmi package installed in sonic-gnmi container.
```

Comment:
Looks like you are introducing new container - Why introduce a new container instead of reusing telemetry? Telemetry container has been used for config purposes by some applications today, can telemetry container be renamed and continue to be used? If a new container is required, what is the backward compatible/migration plan?

Reply:
We did not have an consensus to ahead with the two containers approach one for config and another one for get and subscription in the community call 6/28, please revisit the design and explain why cant we combine all the gNMI operations in one container (today it is telemetry container but can be renamed as gNMI container), this helps to avoid code duplication.

@bhagatyj mentioned in the call today that there is a need for telemetry container restart for new paths addition, if you could explain why cant we do this on-the-fly without container restart, would be helpful to understand the rationale better.
@prvattem Please add if I have missed to capture anything.

Reply:
Agree with the above comments, we should keep a single gnmi server in SONiC, maintaining two gnmi servers in the same NOS unnecessarily divide the code and future contributions. If there are enhancement or capability that needs to be addressed, it should be directed towards the current gnmi server. The gnmi server in telemetry container was introduced 2018, there are active PR contributions in github

Reply:
The gnmi container (Telemetry) already has support for telemetry, config set and config get today. Is there a reason why the existing gnmi server does not fit / cannot be extended for the intended functions ? 
Introducing duplicate containers in the architecture comes with a lot of overhead the community will pay in terms of maintenance.
- The developer community will be divided on which one to contribute to. Does the developer need to ensure both work ? This is an overhead on multiple contributions.
- If it's not mandated that both should work, there will be wide divergence on what features will be supported through which container
- There will be code duplication. Why should the same NOS have two GNMI servers?
If there is something deficient in the existing container kindly list it clearly, and articulate why the existing container cannot be enhanced (if required) to meet the requirements.